components/canvas/ContextMenu.tsx

```diff
@@ -0,0 +1,116 @@
+import { NodeData } from "@/types/NodeData";
+import { getNodeTypeName } from "@/utils/getNodeTypeName";
+import { capitalizeFirstLetter } from "@/utils/stringHelpers";
+import { Menu } from "@equinor/eds-core-react";
+import { RefObject, useState } from "react";
+import { Node, Position } from "reactflow";
+import { MenuItemExandable } from "../MenuItemExandable";
+import styles from "./ContextMenu.module.scss";
+import type { MenuData } from "./hooks/useContextMenu";
+import { useNodeAdd } from "./hooks/useNodeAdd";
+import { getOptionsAddNode } from "./utils/getOptionsAddNode";
+
+type ContextMenuProps = {
+  menuData: MenuData;
+  centerCanvas: () => void;
+  copyToClipBoard?: (target: Node<NodeData>) => Promise<void>;
+  paste?: () => void;
+  onDelete?: (node: Node<NodeData>) => void;
+  onEditNode?: (node: Node<NodeData>) => void;
+  canvasRef?: RefObject<HTMLDivElement>;
+};
+
+export const ContextMenu = ({
+  menuData: { position, node },
+  centerCanvas,
+  copyToClipBoard,
+  paste,
+  onDelete,
+  onEditNode,
+  canvasRef,
+}: ContextMenuProps) => {
+  const [anchorEl, setAnchorEl] = useState<HTMLDivElement | null>(null);
+  const { addNode, isNodeButtonDisabled } = useNodeAdd();
+
+  const isReversedExpandDirection = (fullyExpandedWidth: number) => {
+    const anchorOffsetLeft = anchorEl?.offsetLeft;
+    const canvasEdgeRight = canvasRef?.current?.getBoundingClientRect().right;
+    return !!(
+      canvasEdgeRight &&
+      anchorOffsetLeft &&
+      canvasEdgeRight - anchorOffsetLeft < fullyExpandedWidth
+    );
+  };
+
+  const modifierKey = window.navigator.platform === "MacIntel" ? "⌘" : "Ctrl";
+
+  const renderOptionsAddNode = (node: Node<NodeData>) => {
+    const optionsAddNode = getOptionsAddNode(node);
+    const entries = Object.entries(optionsAddNode);
+    if (entries.length > 0) {
```

Comment:
You could switch to an early return to avoid nesting:

```suggestion
    if (entries.length === 0) return;
```